File: src/lowerMAC/RateAdaptation.cpp
/******************************************************************************
 * WiFiMac                                                                    *
 * This file is part of openWNS (open Wireless Network Simulator)
 * _____________________________________________________________________________
 *
 * Copyright (C) 2004-2007
 * Chair of Communication Networks (ComNets)
 * Kopernikusstr. 16, D-52074 Aachen, Germany
 * phone: ++49-241-80-27910,
 * fax: ++49-241-80-22242
 * email: info@openwns.org
 * www: http://www.openwns.org
 * _____________________________________________________________________________
 *
 * openWNS is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 2 as published by the
 * Free Software Foundation;
 *
 * openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

#include <WIFIMAC/lowerMAC/RateAdaptation.hpp>

using namespace wifimac::lowerMAC;

STATIC_FACTORY_REGISTER_WITH_CREATOR(
    wifimac::lowerMAC::RateAdaptation,
    wns::ldk::FunctionalUnit,
    "wifimac.lowerMAC.RateAdaptation",
    wns::ldk::FUNConfigCreator);

RateAdaptation::RateAdaptation(wns::ldk::fun::FUN* fun, const wns::pyconfig::View& config_) :
    wns::ldk::fu::Plain<RateAdaptation, wns::ldk::EmptyCommand>(fun),
    wns::ldk::Processor<RateAdaptation>(),

    phyUserName(config_.get<std::string>("phyUserName")),
    managerName(config_.get<std::string>("managerName")),
    arqName(config_.get<std::string>("arqName")),
    sinrMIBServiceName(config_.get<std::string>("sinrMIBServiceName")),
    perMIBServiceName(config_.get<std::string>("perMIBServiceName")),
    raForACKFrames(config_.get<bool>("myConfig.raForACKFrames")),
    ackFramesRateId(config_.get<int>("myConfig.ackFramesRateId")),

    config(config_),
    logger(config_.get("logger"))
{
    MESSAGE_SINGLE(NORMAL, this->logger, "created");

    friends.phyUser = NULL;
    friends.manager = NULL;


}


RateAdaptation::~RateAdaptation()
{
}

void RateAdaptation::onFUNCreated()
{
    MESSAGE_SINGLE(NORMAL, this->logger, "onFUNCreated() started");

    friends.phyUser = getFUN()->findFriend<wifimac::convergence::PhyUser*>(phyUserName);
    friends.manager = getFUN()->findFriend<wifimac::lowerMAC::Manager*>(managerName);
    friends.arq = getFUN()->findFriend<wifimac::lowerMAC::ITransmissionCounter*>(arqName);

    sinrMIB = getFUN()->getLayer<dll::Layer2*>()->getManagementService<wifimac::management::SINRInformationBase>(sinrMIBServiceName);
    perMIB = getFUN()->getLayer<dll::Layer2*>()->getManagementService<wifimac::management::PERInformationBase>(perMIBServiceName);

    ackFramesRate = friends.phyUser->getPhyModeProvider()->getPhyMode(ackFramesRateId);
}

void RateAdaptation::processIncoming(const wns::ldk::CompoundPtr& /*compound*/)
{
    // we do exactly nothing with incoming compounds
}

void RateAdaptation::processOutgoing(const wns::ldk::CompoundPtr& compound)
{
    if((not raForACKFrames) and (friends.manager->getFrameType(compound->getCommandPool()) == ACK))
    {
        friends.manager->setPhyMode(compound->getCommandPool(), this->ackFramesRate);

        MESSAGE_BEGIN(NORMAL, logger, m, "Send ACK frame to ");
        m << " rx: " << friends.manager->getReceiverAddress(compound->getCommandPool());
        m << " with " << friends.manager->getPhyMode(compound->getCommandPool());
        MESSAGE_END();
    }
    else
    {
        wns::service::dll::UnicastAddress receiver = friends.manager->getReceiverAddress(compound->getCommandPool());
        size_t numTransmissions = friends.arq->getTransmissionCounter(compound);
        friends.manager->setPhyMode(compound->getCommandPool(), this->getPhyMode(receiver, numTransmissions));

        MESSAGE_BEGIN(NORMAL, logger, m, "Send data frame to ");
        m << " rx: " << friends.manager->getReceiverAddress(compound->getCommandPool());
        m << " with " << friends.manager->getPhyMode(compound->getCommandPool());
        MESSAGE_END();
    }
}

wifimac::convergence::PhyMode
RateAdaptation::getPhyMode(wns::service::dll::UnicastAddress receiver, size_t numTransmissions)
{
    wifimac::convergence::PhyMode pm;

    if(not rateAdaptation.knows(receiver) )
    {
        MESSAGE_BEGIN(NORMAL, logger, m, "No matching RA found for receiver ");
        m << receiver;
        m << ", create new of type " << config.get<std::string>("myConfig.raStrategy.__plugin__");
        MESSAGE_END();

        rateAdaptation.insert(receiver,
                              rateAdaptationStrategies::RateAdaptationStrategyFactory::creator
                              (config.get<std::string>("myConfig.raStrategy.__plugin__"))->create(config.getView("myConfig.raStrategy"),
                                                                                                  perMIB,
                                                                                                  friends.manager,
                                                                                                  friends.phyUser,
                                                                                                  &logger));
    }

    if(sinrMIB->knowsPeerSINR(receiver))
    {
        pm = rateAdaptation.find(receiver)->getPhyMode(receiver, numTransmissions, sinrMIB->getAveragePeerSINR(receiver));
    }
    else
    {
        pm = rateAdaptation.find(receiver)->getPhyMode(receiver, numTransmissions);
    }
    return(pm);
}
wifimac::convergence::PhyMode
RateAdaptation::getPhyMode(const wns::ldk::CompoundPtr& compound)
{
    wns::service::dll::UnicastAddress receiver = friends.manager->getReceiverAddress(compound->getCommandPool());
    size_t numTransmissions = friends.arq->getTransmissionCounter(compound);
    return(this->getPhyMode(receiver, numTransmissions));
}

File: PyConfig/wifimac/lowerMAC/RateAdaptation.py
###############################################################################
# This file is part of openWNS (open Wireless Network Simulator)
# _____________________________________________________________________________
#
# Copyright (C) 2004-2008
# Chair of Communication Networks (ComNets)
# Kopernikusstr. 16, D-52074 Aachen, Germany
# phone: ++49-241-80-27910,
# fax: ++49-241-80-22242
# email: info@openwns.org
# www: http://www.openwns.org
# _____________________________________________________________________________
#
# openWNS is free software; you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License version 2 as published by the
# Free Software Foundation;
#
# openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################

import openwns.FUN
import openwns.pyconfig

import wifimac.Logger
import wifimac.convergence.PhyMode
import wifimac.draftn

class Constant:
    __plugin__ = 'Constant'
    phyMode = None

    def __init__(self, phyMode = None):
        if(phyMode is None):
            self.phyMode = wifimac.convergence.PhyMode.IEEE80211a().getLowest()
        else:
            self.phyMode = phyMode

class PER:
    __plugin__ = 'PER'

    perForGoingDown = 0.25
    """ if the per is above this value, then the next lower phy mode will be used """
    perForGoingUp = 0.01
    """ if the per is below this value, then the next higher phy mode will be used """
    initialPhyMode = None

    def __init__(self, phyMode = None):
        if(phyMode is None):
            self.initialPhyMode = wifimac.convergence.PhyMode.IEEE80211a().getLowest()
        else:
            self.initialPhyMode = phyMode

class ARF:
    __plugin__ = 'ARF'

    arfTimer = 1.0
    initialPhyMode = None
    exponentialBackoff = True
    initialSuccessThreshold = 2
    maxSuccessThreshold = 50

    def __init__(self, phyMode = None):
        if(phyMode is None):
            self.initialPhyMode = wifimac.convergence.PhyMode.IEEE80211a().getLowest()
        else:
            self.initialPhyMode = phyMode

class ARFwithMIMO(ARF):
    __plugin__ = 'ARFwithMIMO'

    def __init__(self, phyMode = None):
        if(phyMode is None):
            self.initialPhyMode = wifimac.draftn.PhyModes().getLowest()
        else:
            self.initialPhyMode = phyMode

class PERwithMIMO(PER):
    __plugin__ = 'PERwithMIMO'

    phyModeIncreaseOnAntennaDecrease = 3
    phyModeDecreaseOnAntennaIncrease = 3

    def __init__(self, phyMode = None):
        if(phyMode is None):
            self.initialPhyMode = wifimac.draftn.PhyModes().getLowest()
        else:
            self.initialPhyMode = phyMode

class SINR(ARF):
    __plugin__ = 'SINR'

    retransmissionLQMReduction = 3.0
    """ Reduce the expected lqm by this value [in dB] for every retransmission of the packet """

class SINRwithMIMO(ARFwithMIMO):
    __plugin__ = 'SINRwithMIMO'

    retransmissionLQMReduction = 3.0
    """ Reduce the expected lqm by this value [in dB] for every retransmission of the packet """

class RateAdaptationConfig(object):
    raStrategy = None
    raForACKFrames = False
    ackFramesRateId = 0

    def __init__(self, **kw):
        self.raStrategy = Constant()
        openwns.pyconfig.attrsetter(self, kw)

class RateAdaptation(openwns.FUN.FunctionalUnit):
    __plugin__ = 'wifimac.lowerMAC.RateAdaptation'

    logger = None
    phyUserName = None
    managerName = None
    arqName = None
    sinrMIBServiceName = None
    perMIBServiceName = None

    myConfig = None

    def __init__(self,
                 functionalUnitName,
                 commandName,
                 phyUserName,
                 managerName,
                 arqName,
                 sinrMIBServiceName,
                 perMIBServiceName,
                 config,
                 parentLogger=None, **kw):
        super(RateAdaptation, self).__init__(functionalUnitName = functionalUnitName, commandName = commandName)
        self.logger = wifimac.Logger.Logger(name = "RateAdaptation", parent = parentLogger)
        self.phyUserName = phyUserName
        self.managerName = managerName
        self.arqName = arqName
        self.sinrMIBServiceName = sinrMIBServiceName
        self.perMIBServiceName = perMIBServiceName
        assert(config.__class__ == RateAdaptationConfig)
        self.myConfig = config
        openwns.pyconfig.attrsetter(self, kw)
